### PR TITLE
FALCON-2231 Changes to support Schedule of user extensions

### PR DESCRIPTION
**Pull request by sandeepSamudrala**

### CODE REVIEW

**Comment by sandeepSamudrala** [comment]
I have dev-tested the same for submit and then scheduling it.

File: prism/src/main/java/org/apache/falcon/resource/proxy/ExtensionManagerProxy.java [hunk 1]
```diff
@@ -191,21 +191,24 @@ public ExtensionInstanceList getInstances(
...
-            if (entities.isEmpty()) {
-                // return failure if the extension job doesn't exist
-                return new APIResult(APIResult.Status.FAILED, "Extension job " + jobName + " doesn't exist.");
-            }
+        ExtensionMetaStore metaStore = ExtensionStore.getMetaStore();
+        ExtensionJobsBean extensionJobsBean = metaStore.getExtensionJobDetails(jobName);
+        if (extensionJobsBean == null) {
+            // return failure if the extension job doesn't exist
+            return new APIResult(APIResult.Status.SUCCEEDED,
+                    "Extension job " + jobName + " doesn't exist. Nothing to delete.");
```
Comment (PracheerAgarwal) [inline prism/src/main/java/org/apache/falcon/resource/proxy/ExtensionManagerProxy.java]: Nothing to delete/schedule?

Comment (sandeepSamudrala) [inline prism/src/main/java/org/apache/falcon/resource/proxy/ExtensionManagerProxy.java]: Ack. Typos. I will correct it.

File: unit/src/main/java/org/apache/falcon/unit/FalconUnitClient.java [hunk 2]
```diff
@@ -334,6 +334,15 @@ public APIResult submitExtensionJob(String extensionName, String jobName, String
     }
 
     @Override
+    public APIResult scheduleExtensionJob(String jobName, String doAsUser) {
+        try {
+            return localExtensionManager.scheduleExtensionJob(jobName, doAsUser);
+        } catch (FalconException | IOException e) {
+            throw new FalconCLIException("Failed to delete the extension job:" + jobName);
```
Comment (PracheerAgarwal) [inline unit/src/main/java/org/apache/falcon/unit/FalconUnitClient.java]: change the error msg to "failed to schedule the extension job"

File: unit/src/main/java/org/apache/falcon/unit/LocalExtensionManager.java [hunk 3]
```diff
@@ -88,7 +104,19 @@ public APIResult submitAndSchedulableExtensionJob(String extensionName, String j
         return new APIResult(APIResult.Status.SUCCEEDED, "Extension job submitted successfully" + jobName);
```
Comment (PracheerAgarwal) [inline unit/src/main/java/org/apache/falcon/unit/LocalExtensionManager.java]: also change the returned message to "scheduled successfully".

Comment (sandeepSamudrala) [inline unit/src/main/java/org/apache/falcon/unit/LocalExtensionManager.java]: Its submit api only.

Comment (PracheerAgarwal) [inline unit/src/main/java/org/apache/falcon/unit/LocalExtensionManager.java]: may be I have commented at wrong place, but messages for both submit and submitAndSchedule was same.

Comment (sandeepSamudrala) [inline unit/src/main/java/org/apache/falcon/unit/LocalExtensionManager.java]: Not related to this change. But I updated the return message for submitAndScheduleExtensionJob Api

Comment (PracheerAgarwal) [inline unit/src/main/java/org/apache/falcon/unit/LocalExtensionManager.java]: thank you @sandeepSamudrala

File: prism/src/main/java/org/apache/falcon/resource/proxy/ExtensionManagerProxy.java [hunk 4]
```diff
@@ -191,21 +191,25 @@ public ExtensionInstanceList getInstances(
...
                               @DefaultValue("") @QueryParam("doAs") String doAsUser) {
         checkIfExtensionServiceIsEnabled();
         checkIfExtensionIsEnabled(ExtensionStore.getMetaStore().getExtensionJobDetails(jobName).getExtensionName());
-        try {
-            List<Entity> entities = getEntityList("", "", "", TAG_PREFIX_EXTENSION_JOB + jobName, "", doAsUser);
-            if (entities.isEmpty()) {
-                // return failure if the extension job doesn't exist
-                return new APIResult(APIResult.Status.FAILED, "Extension job " + jobName + " doesn't exist.");
-            }
+        ExtensionMetaStore metaStore = ExtensionStore.getMetaStore();
```
Comment (pallavi-rao) [inline prism/src/main/java/org/apache/falcon/resource/proxy/ExtensionManagerProxy.java]: The following lines of code, up until getJobEntities, I believe will be repeated for schedule, delete, suspend and resume. See if you can create a common method. I'm fine even if you do that when you implement suspend/resume.

File: unit/src/main/java/org/apache/falcon/unit/LocalExtensionManager.java [hunk 5]
```diff
@@ -65,30 +81,43 @@ public APIResult submitAndSchedulableExtensionJob(String extensionName, String j
...
 
-        return new APIResult(APIResult.Status.SUCCEEDED, "Extension job submitted successfully" + jobName);
+        return new APIResult(APIResult.Status.SUCCEEDED, "Extension job submitted and scheduled successfully"
+                + jobName);
+    }
+
+    APIResult scheduleExtensionJob(String jobName, String doAsUser) throws  FalconException, IOException{
+        ExtensionMetaStore metaStore = ExtensionStore.getMetaStore();
+        ExtensionJobsBean extensionJobsBean = metaStore.getExtensionJobDetails(jobName);
+        SortedMap<EntityType, List<Entity>> entityMap = getJobEntities(extensionJobsBean);
```
Comment (sandeepSamudrala) [inline unit/src/main/java/org/apache/falcon/unit/LocalExtensionManager.java]: Not required its a map and the values are list.. which come as empty.

**Comment by PracheerAgarwal** [comment]
+1

**Comment by PraveenAdlakha** [comment]
+1

**Comment by PraveenAdlakha** [comment]
merging in some time.

**Pull request closed without merging**